File: leetcode/Python/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        sol, new_sol = list(), list() 
        def dfs(visit, pos):
            if pos == len(nums):
                l = []
                for i in range(len(nums)):
                    if visit&(1<<i) > 0:
                        l.append(nums[i])
                sol.append(l)
                return
            dfs(visit, pos+1)
            dfs(visit|(1<<pos), pos+1)
        nums.sort()
        dfs(0,0)
        sol.sort()
        for i in sol:
            if not new_sol:
                new_sol.append(i)
            elif new_sol[-1] != i:
                new_sol.append(i)
        return new_sol

File: leetcode/C++/subsets-ii.cpp
class Solution {
public:
    set<vector<int>> sol;
    void dfs(vector<int>& nums, int visit, int pos){
        if(pos==nums.size()) {
            vector<int> v;
            for(int i=0, p=1; i<nums.size(); ++i, p<<=1){
                if(visit&p) v.push_back(nums[i]);
            }
            sol.insert(v);
            return;
        }
        dfs(nums,visit,pos+1);
        dfs(nums,visit|(1<<pos),pos+1);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        dfs(nums,0,0);
        return vector<vector<int>>(sol.begin(),sol.end());
    }
};